### PR TITLE
ControlMapper refactoring

### PR DESCRIPTION
**Pull request by hrydgard**
This will make it easier to implement multi-input combo mappings as in #14793 . This PR doesn't implement it though, only lays the ground work.

Basically, we flip the entire mapping algorithm on its head - instead of taking an input all the way through, we recompute the related outputs from stored inputs on every input change. This will make combos very easy to implement.

Also, analog virtual keys are now more general and handled by the mapper (including conversion of signed axis to unsigned, as required by PS4 controller triggers on DInput), so VIRTKEY_SPEED_ANALOG is now ~~much~~ simpler.

### CODE REVIEW

**Comment by hrydgard** [comment]
I tried prototyping multi-mapping on top of this, and it comes out pretty well and quite flexible (including more than 2-inputs combos).

Just a little polish left here then I think we can get it in.

**Comment by hrydgard** [comment]
Alright, polish done. Works fine now, and solves an old issue when using both gamepad and keyboard together.

File: Core/ControlMapper.cpp [hunk 1]
```diff
@@ -79,7 +127,7 @@ void ControlMapper::SetPSPAxis(int device, char axis, float value, int stick) {
 	bool ignore = false;
 	if (inDeadZone && lastNonDeadzoneDeviceID_[stick] != device) {
 		// Ignore this event! See issue #15465
-		ignore = true;
+		 ignore = true;
```
Comment (unknownbrackets) [inline Core/ControlMapper.cpp:130]: Looks like a mistake.

-[Unknown]

File: Core/Debugger/WebSocket/InputSubscriber.cpp [hunk 2]
```diff
@@ -192,6 +192,7 @@ void WebSocketInputState::ButtonsPress(DebuggerRequest &req) {
 	}
 	press.button = info->second;
 
+	// TODO: Route into the control mapper's PSPKey function instead.
```
Comment (unknownbrackets) [inline Core/Debugger/WebSocket/InputSubscriber.cpp:195]: I certainly wouldn't want this to have any dependency on `UI` or direct connection to EmuScreen?  What, would there be a global instance of ControlMapper then in Core?

-[Unknown]

File: Core/ControlMapper.cpp [hunk 3]
```diff
@@ -88,49 +136,225 @@ void ControlMapper::SetPSPAxis(int device, char axis, float value, int stick) {
...
+		for (auto &mapping : inputMappings) {
+			if (mapping == changedMapping) {
+				touchedByMapping = true;
+			}
+
+			auto iter = curInput_.find(mapping);
+			if (iter != curInput_.end()) {
+				if (mapping.IsAxis()) {
+					threshold = GetDeviceAxisThreshold(iter->first.deviceId);
+					if (IsUnsignedMapping(vkId)) {
```
Comment (unknownbrackets) [inline Core/ControlMapper.cpp:238]: This really seems to cry out to be a separate method.  Especially with the repeated `value += iter->second;` else cases...

-[Unknown]

File: Core/ControlMapper.cpp [hunk 4]
```diff
@@ -88,49 +136,225 @@ void ControlMapper::SetPSPAxis(int device, char axis, float value, int stick) {
...
+	InputMapping mapping(key.deviceId, key.keyCode);
+
+	if (key.flags & KEY_DOWN) {
+		curInput_[mapping] = 1.0f;
+	} else if (key.flags & KEY_UP) {
+		curInput_[mapping] = 0.0f;
 	}
 
+	// TODO: See if this can be simplified further somehow.
+	bool mappingFound = KeyMap::InputMappingToPspButton(mapping, nullptr);
```
Comment (unknownbrackets) [inline Core/ControlMapper.cpp:330]: Hm, maybe we should at least put it inside `if ((key.flags & KEY_DOWN) && key.keyCode == NKCODE_BACK) {` and put the `if (!mappingFound || key.deviceId == DEVICE_ID_DEFAULT) {` inside that too.

-[Unknown]

File: Core/HLE/sceCtrl.h [hunk 5]
```diff
@@ -70,7 +70,8 @@ void __CtrlShutdown();
 void __CtrlButtonDown(u32 buttonBit);
 // Call this whenever a button is released.  Similar to __CtrlButtonDown().
 void __CtrlButtonUp(u32 buttonBit);
-// To be used by the new mapping code.
+// Sets the full state. Used by the new control mapper. The above two functions
+// should go away over time.
```
[the remark this answers is not in the document]
Comment (hrydgard) [reply Core/HLE/sceCtrl.h:74]: I'll just replace it all with CtrlUpdateButtons and change the comments. Agree that it's ended up a bit confusing after a couple of iterations.

File: Core/Debugger/WebSocket/InputSubscriber.cpp (hunk 2 above)
Comment (hrydgard) [reply Core/Debugger/WebSocket/InputSubscriber.cpp:195]: Yeah I've changed my mind on this already. I'll get rid of these comments.

File: Core/ControlMapper.cpp (hunk 3 above)
Comment (hrydgard) [reply Core/ControlMapper.cpp:238]: Will do.

File: Core/ControlMapper.cpp (hunk 4 above)
Comment (hrydgard) [reply Core/ControlMapper.cpp:330]: Right, that'll save a little bit indeed. Fixing.

**Review by unknownbrackets — APPROVED** [review]

**Pull request merged**